// heroku/dist/commands/webhooks/events/index.js
import { flags } from '@heroku-cli/command';
import { color, hux } from '@heroku/heroku-cli-util';
import { ux } from '@oclif/core/ux';
import BaseCommand from '../../../lib/webhooks/base.js';
export default class EventsIndex extends BaseCommand {
    static description = 'list webhook events on an app';
    static examples = [
        `${color.command('heroku webhooks:events')}`,
    ];
    static flags = {
        app: flags.app(),
        pipeline: flags.pipeline({ char: 'p', description: 'pipeline on which to list', hidden: true }),
        remote: flags.remote(),
    };
    async run() {
        const { flags } = await this.parse(EventsIndex);
        const { display, path } = this.webhookType(flags);
        ux.warn('heroku webhooks:event is deprecated, please use heroku webhooks:deliveries');
        const { body: events } = await this.webhooksClient.get(`${path}/webhook-events`);
        if (events.length === 0) {
            this.log(`${display} has no events`);
        }
        else {
            events.sort((a, b) => Date.parse(a.created_at) - Date.parse(b.created_at));
            const printLine = (...args) => this.log(...args);
            /* eslint-disable perfectionist/sort-objects */
            hux.table(events, {
                id: {
                    header: 'Event ID',
                },
                resource: {
                    header: 'Resource',
                    get: (w) => w.payload.resource,
                },
                action: {
                    header: 'Action',
                    get: (w) => w.payload.action,
                },
                published_at: {
                    header: 'Published At',
                    get: (w) => w.payload.published_at,
                },
            }, {
                printLine,
            });
            /* eslint-enable perfectionist/sort-objects */
        }
    }
}
