// heroku/dist/lib/pg/backups.js
import { createRequire as _createRequire } from "module";
const __require = _createRequire(import.meta.url);
import { color, utils } from '@heroku/heroku-cli-util';
import { ux } from '@oclif/core/ux';
import tsheredoc from 'tsheredoc';
const bytes = __require("bytes");
const heredoc = tsheredoc.default;
function prefix(transfer) {
    if (transfer.from_type === 'pg_dump') {
        if (transfer.to_type === 'pg_restore') {
            return 'c';
        }
        return transfer.schedule ? 'a' : 'b';
        // eslint-disable-next-line no-else-return
    }
    else {
        if (transfer.to_type === 'pg_restore') {
            return 'r';
        }
        return 'b';
    }
}
class Backups {
    app;
    heroku;
    logsAlreadyShown = new Set();
    constructor(app, heroku) {
        this.app = app;
        this.heroku = heroku;
    }
    displayLogs(logs) {
        for (const log of logs) {
            if (this.logsAlreadyShown.has(log.created_at + log.message)) {
                continue;
            }
            this.logsAlreadyShown.add(log.created_at + log.message);
            ux.stdout(`${log.created_at} ${log.message}`);
        }
    }
    filesize(size, opts = {}) {
        Object.assign(opts, {
            decimalPlaces: 2,
            fixedDecimals: true,
        });
        return bytes(size, opts) || '';
    }
    name(transfer) {
        const oldPGBName = transfer.options?.pgbackups_name;
        if (oldPGBName)
            return `o${oldPGBName}`;
        return `${prefix(transfer)}${(transfer.num || '').toString().padStart(3, '0')}`;
    }
    async num(name) {
        let m = name.match(/^[abcr](\d+)$/);
        if (m)
            return Number.parseInt(m[1], 10);
        m = name.match(/^o[ab]\d+$/);
        if (m) {
            const { body: transfers } = await this.heroku.get(`/client/v11/apps/${this.app}/transfers`, { hostname: utils.pg.host() });
            const transfer = transfers.find(t => this.name(t) === name);
            if (transfer)
                return transfer.num;
        }
    }
    async *poll(transferID, interval, verbose, appId) {
        const tty = process.env.TERM !== 'dumb' && process.stderr.isTTY;
        let backup = {};
        let failures = 0;
        const quietUrl = `/client/v11/apps/${appId ?? this.app}/transfers/${transferID}`;
        const verboseUrl = quietUrl + '?verbose=true';
        const url = verbose ? verboseUrl : quietUrl;
        while (failures < 21) {
            try {
                ({ body: backup } = await this.heroku.get(url, { hostname: utils.pg.host() }));
            }
            catch (error) {
                if (failures++ > 20) {
                    throw error;
                }
            }
            if (verbose) {
                this.displayLogs(backup.logs);
            }
            else if (tty) {
                const msg = backup.started_at ? this.filesize(backup.processed_bytes) : 'pending';
                const log = backup.logs?.pop();
                if (log) {
                    ux.action.status = `${msg}\n${log.created_at + ' ' + log.message}`;
                }
                else {
                    ux.action.status = msg;
                }
            }
            if (backup?.finished_at) {
                if (backup.succeeded) {
                    yield true;
                    break;
                }
                // logs is undefined unless verbose=true is passed
                ({ body: backup } = await this.heroku.get(verboseUrl, { hostname: utils.pg.host() }));
                throw new Error(heredoc(`
          An error occurred and the backup did not finish.

          ${backup.logs.slice(-5).map(l => l.message).join('\n')}

          Run ${color.code('heroku pg:backups:info ' + this.name(backup))} for more details.`));
            }
            yield new Promise(resolve => {
                setTimeout(resolve, interval * 1000);
            });
        }
    }
    status(transfer) {
        const { finished_at, processed_bytes, started_at, succeeded, warnings } = transfer;
        if (finished_at && succeeded) {
            if (warnings > 0) {
                return `Finished with ${warnings} warnings`;
            }
            return `Completed ${finished_at}`;
        }
        if (finished_at) {
            return `Failed ${finished_at}`;
        }
        if (started_at) {
            return `Running (processed ${this.filesize(processed_bytes)})`;
        }
        return 'Pending';
    }
    async wait(action, transferID, interval, verbose, app) {
        if (verbose) {
            ux.stdout(`${action}...`);
        }
        ux.action.start(action);
        try {
            for await (const backupSucceeded of this.poll(transferID, interval, verbose, app || this.app)) {
                if (backupSucceeded) {
                    ux.action.stop();
                    break;
                }
            }
        }
        catch (error) {
            ux.action.stop('!');
            ux.error(error);
        }
    }
}
function factory(app, heroku) {
    return new Backups(app, heroku);
}
export default factory;
