// heroku/dist/lib/pg/setter.js
import { Command, flags } from '@heroku-cli/command';
import { ux } from '@oclif/core/ux';
import { utils } from '@heroku/heroku-cli-util';
import { essentialPlan } from './util.js';
export class PGSettingsCommand extends Command {
    static topic = 'pg';
    static flags = {
        app: flags.app({ required: true }),
        remote: flags.remote(),
    };
    async run() {
        const { flags, args } = await this.parse();
        const { app } = flags;
        const { value, database } = args;
        const dbResolver = new utils.pg.DatabaseResolver(this.heroku);
        const { addon: db } = await dbResolver.getAttachment(app, database);
        if (essentialPlan(db)) {
            ux.error('You can\'t perform this operation on Essential-tier databases.');
        }
        if (value) {
            const { body: settings } = await this.heroku.patch(`/postgres/v0/databases/${db.id}/config`, {
                hostname: utils.pg.host(),
                body: { [this.settingKey]: this.convertValue(value) },
            });
            const setting = settings[this.settingKey];
            ux.stdout(`${this.settingKey.replace(/_/g, '-')} has been set to ${setting.value} for ${db.name}.`);
            ux.stdout(this.explain(setting));
        }
        else {
            const { body: settings } = await this.heroku.get(`/postgres/v0/databases/${db.id}/config`, { hostname: utils.pg.host() });
            const setting = settings[this.settingKey];
            ux.stdout(`${this.settingKey.replace(/_/g, '-')} is set to ${setting.value} for ${db.name}.`);
            ux.stdout(this.explain(setting));
        }
    }
}
export const booleanConverter = (value) => {
    switch (value) {
        case 'true':
        case 'TRUE':
        case 'ON':
        case 'on': {
            return true;
        }
        case 'false':
        case 'FALSE':
        case 'OFF':
        case 'off':
        case null: {
            return false;
        }
        default: {
            throw new TypeError('Invalid value. Valid options are: a boolean value');
        }
    }
};
export const numericConverter = (value) => {
    const n = Number(value);
    if (!Number.isFinite(n)) {
        throw new TypeError('Invalid value. Valid options are: a numeric value');
    }
    return n;
};
