// heroku/dist/lib/webhooks/base.js
import * as color from '@heroku/heroku-cli-util/color';
import { APIClient, Command } from '@heroku-cli/command';
export default class extends Command {
    webhooksClient;
    constructor(argv, config) {
        super(argv, config);
        const client = new APIClient(this.config, {});
        client.defaults.headers = {
            ...this.heroku.defaults.headers,
            Accept: 'application/vnd.heroku+json; version=3.webhooks',
            authorization: `Basic ${Buffer.from(':' + this.heroku.auth).toString('base64')}`,
        };
        this.webhooksClient = client;
    }
    webhookType(context) {
        if (context.pipeline) {
            return {
                display: context.pipeline,
                path: `/pipelines/${context.pipeline}`,
            };
        }
        if (context.app) {
            return {
                display: color.app(context.app),
                path: `/apps/${context.app}`,
            };
        }
        return this.error('No app specified');
    }
}
