// heroku/dist/commands/labs/enable.js
import { Command, flags } from '@heroku-cli/command';
import * as color from '@heroku/heroku-cli-util/color';
import { Args, ux } from '@oclif/core';
export default class LabsEnable extends Command {
    static args = {
        feature: Args.string({ description: 'unique identifier or name of the account feature', required: true }),
    };
    static description = 'enables an experimental feature';
    static flags = {
        app: flags.app({ required: false }),
        remote: flags.remote(),
    };
    static topic = 'labs';
    async run() {
        const { args, flags } = await this.parse(LabsEnable);
        const { feature } = args;
        let target;
        let request;
        try {
            await this.heroku.get(`/account/features/${feature}`);
            request = enableFeature(this.heroku, feature);
            const targetResponse = await this.heroku.get('/account');
            target = color.user(targetResponse.body.email || '');
        }
        catch (error) {
            if (error.http.statusCode !== 404)
                throw error;
            // might be an app feature
            if (!flags.app)
                throw error;
            await this.heroku.get(`/apps/${flags.app}/features/${feature}`);
            request = enableFeature(this.heroku, feature, flags.app);
            target = color.app(flags.app);
        }
        ux.action.start(`Enabling ${color.name(feature)} for ${target}`);
        await request;
        ux.action.stop();
    }
}
async function enableFeature(heroku, feature, app) {
    return heroku.patch(app ? `/apps/${app}/features/${feature}` : `/account/features/${feature}`, {
        body: { enabled: true },
    });
}
