// heroku/dist/commands/telemetry/index.js
import { Command, flags as Flags } from '@heroku-cli/command';
import { color, hux } from '@heroku/heroku-cli-util';
import { ux } from '@oclif/core/ux';
export default class Index extends Command {
    static description = 'list telemetry drains';
    static example = `${color.command('heroku telemetry')}`;
    static flags = {
        app: Flags.string({ char: 'a', description: 'filter by app name' }),
        space: Flags.string({ char: 's', description: 'filter by space name', exactlyOne: ['app', 'space'] }),
    };
    static topic = 'telemetry';
    display(telemetryDrains, owner) {
        if (telemetryDrains.length === 0) {
            ux.stdout(`There are no telemetry drains in ${owner}`);
        }
        else {
            hux.styledHeader(`${owner} Telemetry Drains`);
            /* eslint-disable perfectionist/sort-objects */
            hux.table(telemetryDrains, {
                ID: { get: telemetryDrain => telemetryDrain.id },
                Signals: { get: telemetryDrain => telemetryDrain.signals },
                Endpoint: { get: telemetryDrain => telemetryDrain.exporter.endpoint },
            });
            /* eslint-enable perfectionist/sort-objects */
        }
    }
    async run() {
        const { flags } = await this.parse(Index);
        const { app, space } = flags;
        let drains = [];
        if (app) {
            const { body: appTelemetryDrains } = await this.heroku.get(`/apps/${app}/telemetry-drains`, {
                headers: {
                    Accept: 'application/vnd.heroku+json; version=3.sdk',
                },
            });
            drains = appTelemetryDrains;
        }
        else if (space) {
            const { body: spaceTelemetryDrains } = await this.heroku.get(`/spaces/${space}/telemetry-drains`, {
                headers: {
                    Accept: 'application/vnd.heroku+json; version=3.sdk',
                },
            });
            drains = spaceTelemetryDrains;
        }
        this.display(drains, app || space);
    }
}
