// heroku/dist/commands/pg/settings/data-connector-details-logs.js
import { flags } from '@heroku-cli/command';
import { Args } from '@oclif/core';
import tsheredoc from 'tsheredoc';
import { booleanConverter, PGSettingsCommand } from '../../../lib/pg/setter.js';
import { nls } from '../../../nls.js';
const heredoc = tsheredoc.default;
export default class DataConnectorDetailsLogs extends PGSettingsCommand {
    static aliases = ['pg:settings:explain-data-connector-details'];
    static description = heredoc(`
  displays stats on replication slots on your database, the default value is "off"
  `);
    static flags = {
        app: flags.app({ required: true }),
        remote: flags.remote(),
    };
    static args = {
        database: Args.string({ description: `${nls('pg:database:arg:description')} ${nls('pg:database:arg:description:default:suffix')}` }),
        value: Args.string({ description: 'boolean indicating if data replication slot details get logged' }),
    };
    settingKey = 'data_connector_details_logs';
    convertValue(val) {
        return booleanConverter(val);
    }
    explain(setting) {
        if (setting?.value) {
            return 'Data replication slot details will be logged.';
        }
        return 'Data replication slot details will no longer be logged.';
    }
}
