// heroku/dist/lib/addons/util.js
import * as color from '@heroku/heroku-cli-util/color';
import printf from 'printf';
import ConfirmCommand from '../confirmCommand.js';
export const trapConfirmationRequired = async (app, confirm, fn) => {
    try {
        return await fn(confirm);
    }
    catch (error) {
        if (!isHttpError(error) || error.body?.id !== 'confirmation_required') {
            throw error;
        }
        await new ConfirmCommand().confirm(app, confirm, error.body.message);
        return fn(app);
    }
};
function isHttpError(error) {
    return Boolean(error) && error instanceof Error && Reflect.has(error, 'body');
}
// This function assumes that price.cents will reflect price per month.
// If the API returns any unit other than month
// this function will need to be updated.
export const formatPrice = function ({ hourly, price }) {
    if (!price)
        return;
    if (price.contract)
        return 'contract';
    if (price.metered)
        return 'metered';
    if (price.cents === 0)
        return 'free';
    // we are using a standardized 720 hours/month
    if (hourly)
        return `~$${((price.cents / 100) / 720).toFixed(3)}/hour`;
    const fmt = price.cents % 100 === 0 ? '$%.0f/%s' : '$%.02f/%s';
    return printf(fmt, price.cents / 100, price.unit);
};
export const formatPriceText = function (price) {
    const priceHourly = formatPrice({ hourly: true, price });
    const priceMonthly = formatPrice({ hourly: false, price });
    if (!priceHourly)
        return '';
    if (priceHourly === 'free' || priceHourly === 'contract' || priceHourly === 'metered')
        return `${color.info(priceHourly)}`;
    return `${color.info(priceHourly)} (max ${priceMonthly})`;
};
export const grandfatheredPrice = function (addon) {
    const price = addon.plan?.price;
    return {
        ...price,
        cents: addon.billed_price?.cents,
        contract: addon.billed_price?.contract,
    };
};
export const formatState = function (state) {
    switch (state) {
        case 'provisioned': {
            state = 'created';
            break;
        }
        case 'provisioning': {
            state = 'creating';
            break;
        }
        case 'deprovisioning': {
            state = 'destroying';
            break;
        }
        case 'deprovisioned': {
            state = 'errored';
            break;
        }
        default: {
            state = '';
            break;
        }
    }
    return state;
};
