// @heroku/heroku-cli-util/dist/ux/colors.js
import ansis from 'ansis';
/**
 * Color constants for Heroku CLI output
 * Most colors use ANSI256 codes for better compatibility with terminals that don't support TrueColor.
 * Magenta, red, and cyan are kept as hex values for precise color matching.
 */
const COLORS = {
    // Blue tones (ANSI256: 117)
    BLUE: 117,
    // Command tones (dark gray background, white foreground)
    // ANSI256: 237 (background), 255 (foreground)
    CODE_BG: 237,
    CODE_FG: 255,
    // Cyan tones (kept as hex for precise color matching)
    CYAN: '#50D3D5',
    CYAN_LIGHT: '#8FF5F7',
    // Gold tones (ANSI256: 220, approx. #FFD700)
    GOLD: 220,
    // Gray tones (ANSI256: 248)
    GRAY: 248,
    // Green tones (ANSI256: 40)
    GREEN: 40,
    // Magenta tones (kept as hex for precise color matching)
    MAGENTA: '#FF22DD',
    // Orange tones (ANSI256: 214)
    ORANGE: 214,
    // Pink tones (ANSI256: 212, approx. #FF87D7 - closest match to original #FF8DD3)
    PINK: 212,
    // Purple tones (ANSI256: 147 - closest match to original #ACADFF)
    PURPLE: 147,
    // Red tones (kept as hex for precise color matching)
    RED: '#FF8787',
    // Teal tones (ANSI256: 43)
    TEAL: 43,
    // Yellow tones (ANSI256: 185)
    YELLOW: 185,
};
/**
 * Color definitions for Heroku CLI output
 * Each color has a specific purpose as defined in the design system
 * Colors use ANSI256 codes where possible, with hex values for magenta, red, and cyan
 */
// Helper function to apply color based on type (number = ANSI256, string = hex)
const colorize = (color) => typeof color === 'number' ? ansis.fg(color) : ansis.hex(color);
const bgColorize = (color) => typeof color === 'number' ? ansis.bg(color) : ansis.bgHex(color);
// Check if terminal supports at least ANSI256 (level >= 2)
// Level values: 0=no color, 1=ANSI16, 2=ANSI256, 3=TrueColor
const supportsAnsi256 = ansis.level >= 2;
// Helper function for purple color that falls back to ANSI16 magenta when only ANSI16 is supported
const purpleColorize = () => supportsAnsi256 ? ansis.fg(COLORS.PURPLE) : ansis.magenta;
const THEME_ENV = 'HEROKU_THEME';
/**
 * Resolves active theme from HEROKU_THEME; defaults to 'heroku'.
 * @returns The active theme name.
 */
export function getTheme() {
    const v = process.env[THEME_ENV]?.toLowerCase().trim();
    if (v === 'simple')
        return 'simple';
    return 'heroku';
}
const herokuTheme = {
    addon: (text) => colorize(COLORS.YELLOW)(text),
    app: (text) => purpleColorize()(`${supportsAnsi256 ? '⬢ ' : ''}${text}`),
    attachment: (text) => colorize(COLORS.GOLD)(text),
    code: (text) => bgColorize(COLORS.CODE_BG).fg(COLORS.CODE_FG).bold(`${text}`),
    command: (text) => bgColorize(COLORS.CODE_BG).fg(COLORS.CODE_FG).bold(` $ ${text} `),
    datastore: (text) => colorize(COLORS.YELLOW)(`${supportsAnsi256 ? '⛁ ' : ''}${text}`),
    failure: (text) => colorize(COLORS.RED)(text),
    inactive: (text) => colorize(COLORS.GRAY)(text),
    info: (text) => colorize(COLORS.TEAL)(text),
    label: (text) => ansis.bold(text),
    name: (text) => colorize(COLORS.PINK)(text),
    pipeline: (text) => colorize(COLORS.MAGENTA)(text),
    space: (text) => colorize(COLORS.BLUE)(`${supportsAnsi256 ? '⬡ ' : ''}${text}`),
    success: (text) => colorize(COLORS.GREEN)(text),
    team: (text) => colorize(COLORS.CYAN_LIGHT)(text),
    user: (text) => colorize(COLORS.CYAN)(text),
    warning: (text) => colorize(COLORS.ORANGE)(text),
};
/** Simple theme: ANSI 8 colors only, no symbols. */
const simpleTheme = {
    addon: (text) => ansis.yellow(text),
    app: (text) => ansis.magenta(text),
    attachment: (text) => ansis.yellow(text),
    code: (text) => ansis.bold(text),
    command: (text) => ansis.bold(` $ ${text} `),
    datastore: (text) => ansis.yellow(text),
    failure: (text) => ansis.red(text),
    inactive: (text) => ansis.gray(text),
    info: (text) => ansis.cyan(text),
    label: (text) => ansis.bold(text),
    name: (text) => ansis.magenta(text),
    pipeline: (text) => ansis.magenta(text),
    space: (text) => ansis.cyan(text),
    success: (text) => ansis.green(text),
    team: (text) => ansis.cyan(text),
    user: (text) => ansis.cyan(text),
    warning: (text) => ansis.yellow(text),
};
const activeTheme = () => (getTheme() === 'simple' ? simpleTheme : herokuTheme);
// Colors for entities on the Heroku platform
export const app = (text) => activeTheme().app(text);
export const pipeline = (text) => activeTheme().pipeline(text);
export const space = (text) => activeTheme().space(text);
export const datastore = (text) => activeTheme().datastore(text);
export const addon = (text) => activeTheme().addon(text);
export const attachment = (text) => activeTheme().attachment(text);
export const name = (text) => activeTheme().name(text);
// Status colors
export const success = (text) => activeTheme().success(text);
export const enabled = success;
export const failure = (text) => activeTheme().failure(text);
export const error = failure;
export const warning = (text) => activeTheme().warning(text);
// User/Team colors
export const team = (text) => activeTheme().team(text);
export const user = (text) => activeTheme().user(text);
// General purpose colors
export const label = (text) => activeTheme().label(text);
export const info = (text) => activeTheme().info(text);
export const inactive = (text) => activeTheme().inactive(text);
export const disabled = inactive;
export const command = (text) => activeTheme().command(text);
export const code = (text) => activeTheme().code(text);
export const snippet = code;
/**
 * Color palette for reference
 * Shows ANSI256 codes for most colors, hex values for magenta, red, and cyan
 */
export const colorPalette = {
    addon: { name: 'yellow', style: 'normal', value: COLORS.YELLOW },
    app: { name: 'purple', style: 'normal', value: COLORS.PURPLE },
    attachment: { name: 'yellow', style: 'normal', value: COLORS.YELLOW },
    code: { name: 'white on dark gray', style: 'bold', value: COLORS.CODE_FG },
    command: { name: 'white on dark gray', style: 'bold', value: COLORS.CODE_FG },
    datastore: { name: 'yellow', style: 'normal', value: COLORS.YELLOW },
    failure: { name: 'red', style: 'normal', value: COLORS.RED },
    gold: { name: 'gold', style: 'normal', value: COLORS.GOLD },
    inactive: { name: 'gray', style: 'normal', value: COLORS.GRAY },
    info: { name: 'teal', style: 'normal', value: COLORS.TEAL },
    label: { name: 'bright white/black', style: 'bold', value: 'default' },
    name: { name: 'magenta', style: 'normal', value: COLORS.MAGENTA },
    pipeline: { name: 'purple', style: 'normal', value: COLORS.PURPLE },
    snippet: { name: 'white on dark gray', style: 'normal', value: COLORS.CODE_FG },
    space: { name: 'blue', style: 'normal', value: COLORS.BLUE },
    success: { name: 'green', style: 'normal', value: COLORS.GREEN },
    team: { name: 'light cyan', style: 'normal', value: COLORS.CYAN_LIGHT },
    user: { name: 'cyan', style: 'normal', value: COLORS.CYAN },
    warning: { name: 'orange', style: 'normal', value: COLORS.ORANGE },
};
/**
 * Export color constants for external use
 */
export { COLORS };
/**
 * Re-export everything from ansis as a passthrough
 * This gives access to all ansis functionality while adding our custom colors
 */
export * from 'ansis';
export { default as ansi } from 'ansis';
