// heroku/dist/lib/pipelines/setup/create-apps.js
import { ux } from '@oclif/core/ux';
import { createAppSetup } from '../../api.js';
export default function createApps(heroku, archiveURL, pipeline, pipelineName, stagingAppName, organization) {
    const prodAppSetupPromise = createApp(heroku, {
        archiveURL,
        name: pipelineName,
        organization,
        pipeline,
        stage: 'production',
    });
    const stagingAppSetupPromise = createApp(heroku, {
        archiveURL,
        name: stagingAppName,
        organization,
        pipeline,
        stage: 'staging',
    });
    const promises = [prodAppSetupPromise, stagingAppSetupPromise];
    return Promise.all(promises).then(appSetups => appSetups, error => {
        ux.error(error, { exit: 1 });
    });
}
function createApp(heroku, { archiveURL, name, organization, pipeline, stage }) {
    const params = {
        app: { name },
        pipeline_coupling: {
            pipeline: pipeline.id,
            stage,
        },
        source_blob: { url: archiveURL },
    };
    if (organization) {
        params.app.organization = organization;
    }
    else {
        params.app.personal = true;
    }
    return createAppSetup(heroku, params).then(setup => setup);
}
